doc/adding_lints.md

```diff
@@ -0,0 +1,366 @@
+        if !is_foo_fn(fn_kind) { return; }
+        span_help_and_lint(
+            cx,
+            FOO_FUNCTIONS,
+            span,
+            "function named `foo`",
+            "consider using a more meaningful name"
+        );
+    }
+}
+```
+
+We separate the lint conditional from the lint emissions because it makes the
+code a bit easier to read. In some cases this separation would also allow to
+write some unit tests (as opposed to UI tests) for the separate function.
+
+In our example, `is_foo_fn` looks like:
+
+```rust
+// use statements, impl EarlyLintPass, check_fn, ..
+
+fn is_foo_fn(fn_kind: FnKind<'_>) -> bool {
+    match fn_kind {
+        FnKind::ItemFn(ident, ..) | FnKind::Method(ident, ..) => {
+            ident.name == "foo"
+        },
+        FnKind::Closure(..) => false
+    }
+}
+```
+
+Now you'll want to also run the full test suite with `cargo test`. Apart from
+running all other UI tests, this ensures that our lint implementation is not
+violating any Clippy lints itself.
+
+If you are still following the example, you'll see that the `FooFunctionsPass`
+violates a Clippy lint. So we are going to rename that struct to just `Pass`:
+
+```rust
+#[derive(Copy, Clone)]
+pub struct Pass;
+
+impl LintPass for Pass { /* .. */ }
+```
+
+That should be it for the lint implementation. Running `cargo test` should now
+pass and we can finish up our work by adding some documentation.
+
+### Documentation
+
+The final thing before submitting our PR is to add some documentation to our
+lint declaration.
+
+Please document your lint with a doc comment akin to the following:
+
+```rust
+/// **What it does:** Checks for ... (describe what the lint matches).
+///
+/// **Why is this bad?** Supply the reason for linting the code.
+///
+/// **Known problems:** None. (Or describe where it could go wrong.)
+///
+/// **Example:**
+///
+/// ```rust,ignore
+/// // Bad
+/// Insert a short example of code that triggers the lint
+///
+/// // Good
+/// Insert a short example of improved code that doesn't trigger the lint
+/// ```
+declare_clippy_lint! { /* ... */ }
+```
+
+Once your lint is merged, this documentation will show up in the [lint
+list][lint_list].
+
+### Debugging
+
+If you want to debug parts of your lint implementation, you can use the `dbg!`
+macro anywhere in your code. Running the tests should then include the debug
+output in the `stdout` part.
+
+### PR Checklist
+
+TODO: Prose
+
+- [ ] Followed [lint naming conventions][lint_naming]
+- [ ] Added passing UI tests (including committed `.stderr` file)
+- [ ] `cargo test` passes locally
+- [ ] Added lint documentation
+
+### Cheatsheet
+
+Here are some pointers to things you are likely going to need for every lint:
+
+* [The `if_chain` macro][if_chain]
+* [`in_macro`][in_macro] and [`in_external_macro`][in_external_macro]
+* [`Span`][span]
+* [Clippy diagnostics][diagnostics]
+* [`Applicability`][applicability]
```

Comment:
Maybe also mention the `utils` module here. There are many helpful convenience functions ready to use. (`snippet`, `Sugg`, `is_expn_of`, ...)

Reply:
Done in https://github.com/rust-lang/rust-clippy/pull/3824/commits/31b62841e0252e3d33e10e9b3f39b9b7e2c8f63d